# **DATASET AUDIT - RETAIL SALES INTELLIGENCE ENGINE**

In [1]:
import pandas as pd

In [3]:
df = pd.read_csv(r"E:\M90\PROJECTS\MY PROJECTS\PROJECT 3 — Retail Store Sales Intelligence & Forecasting Engine\DATASET\sales data.csv")

C:\Users\abhin\AppData\Local\Temp\ipykernel_20756\2905592497.py:1: DtypeWarning: Columns (23) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(r"E:\M90\PROJECTS\MY PROJECTS\PROJECT 3 — Retail Store Sales Intelligence & Forecasting Engine\DATASET\sales data.csv")


In [4]:
print (df.shape)

(128975, 24)


**SHAPE OF DF = (128975, 24)**

In [5]:
print (df.columns)

Index(['index', 'Order ID', 'Date', 'Status', 'Fulfilment', 'Sales Channel ',
       'ship-service-level', 'Style', 'SKU', 'Category', 'Size', 'ASIN',
       'Courier Status', 'Qty', 'currency', 'Amount', 'ship-city',
       'ship-state', 'ship-postal-code', 'ship-country', 'promotion-ids',
       'B2B', 'fulfilled-by', 'Unnamed: 22'],
      dtype='object')


In [9]:
print (df.dtypes)

index                   int64
Order ID               object
Date                   object
Status                 object
Fulfilment             object
Sales Channel          object
ship-service-level     object
Style                  object
SKU                    object
Category               object
Size                   object
ASIN                   object
Courier Status         object
Qty                     int64
currency               object
Amount                float64
ship-city              object
ship-state             object
ship-postal-code      float64
ship-country           object
promotion-ids          object
B2B                      bool
fulfilled-by           object
Unnamed: 22            object
dtype: object


In [12]:
print (df.head(20))

    index             Order ID      Date                        Status  \
0       0  405-8078784-5731545  04-30-22                     Cancelled   
1       1  171-9198151-1101146  04-30-22  Shipped - Delivered to Buyer   
2       2  404-0687676-7273146  04-30-22                       Shipped   
3       3  403-9615377-8133951  04-30-22                     Cancelled   
4       4  407-1069790-7240320  04-30-22                       Shipped   
5       5  404-1490984-4578765  04-30-22                       Shipped   
6       6  408-5748499-6859555  04-30-22                       Shipped   
7       7  406-7807733-3785945  04-30-22  Shipped - Delivered to Buyer   
8       8  407-5443024-5233168  04-30-22                     Cancelled   
9       9  402-4393761-0311520  04-30-22                       Shipped   
10     10  407-5633625-6970741  04-30-22                       Shipped   
11     11  171-4638481-6326716  04-30-22                       Shipped   
12     12  405-5513694-8146768  04-30-

**INITIAL OBSERVATIONS**  
Ship city - Not unique formats  
Currency - Empty cells  
Courier Status - Data missing  
Amount - Missing Content  

In [13]:
missing_values_per_column = df.isnull().sum()
print (missing_values_per_column)

index                     0
Order ID                  0
Date                      0
Status                    0
Fulfilment                0
Sales Channel             0
ship-service-level        0
Style                     0
SKU                       0
Category                  0
Size                      0
ASIN                      0
Courier Status         6872
Qty                       0
currency               7795
Amount                 7795
ship-city                33
ship-state               33
ship-postal-code         33
ship-country             33
promotion-ids         49153
B2B                       0
fulfilled-by          89698
Unnamed: 22           49050
dtype: int64


In [16]:
total_missing_values = missing_values_per_column.sum()
print (total_missing_values)

210495


In [17]:
total_rows = len(df)

In [25]:
percentage_of_missing_values = (missing_values_per_column/total_rows)*100
print(percentage_of_missing_values)

index                  0.000000
Order ID               0.000000
Date                   0.000000
Status                 0.000000
Fulfilment             0.000000
Sales Channel          0.000000
ship-service-level     0.000000
Style                  0.000000
SKU                    0.000000
Category               0.000000
Size                   0.000000
ASIN                   0.000000
Courier Status         5.328164
Qty                    0.000000
currency               6.043807
Amount                 6.043807
ship-city              0.025586
ship-state             0.025586
ship-postal-code       0.025586
ship-country           0.025586
promotion-ids         38.110487
B2B                    0.000000
fulfilled-by          69.546811
Unnamed: 22           38.030626
dtype: float64


**MISSING VALUE OBSERVATIONS**  
total_missing_values = 210495  
  
 missing values per column  
 index                    0  
Order ID                  0  
Date                      0  
Status                    0  
Fulfilment                0  
Sales Channel             0  
ship-service-level        0  
Style                     0  
SKU                       0  
Category                  0  
Size                      0  
ASIN                      0  
Courier Status         6872  
Qty                       0  
currency               7795  
Amount                 7795  
ship-city                33  
ship-state               33  
ship-postal-code         33  
ship-country             33  
promotion-ids         49153  
B2B                       0  
fulfilled-by          89698  
Unnamed: 22           49050  
dtype: int64  


In [38]:
duplicates = df.duplicated().any()
print (duplicates)

False


In [28]:
duplicate_rows = df[duplicates]
print (duplicate_rows)

Empty DataFrame
Columns: [index, Order ID, Date, Status, Fulfilment, Sales Channel , ship-service-level, Style, SKU, Category, Size, ASIN, Courier Status, Qty, currency, Amount, ship-city, ship-state, ship-postal-code, ship-country, promotion-ids, B2B, fulfilled-by, Unnamed: 22]
Index: []

[0 rows x 24 columns]


In [32]:
duplicate_order_id = df["Order ID"].duplicated().any()
print (duplicate_order_id)

True


In [37]:
duplicate_orderid_rows = df[df['Order ID'].duplicated(keep=False)]
print (duplicate_orderid_rows['Order ID'].head(15443))

37        403-4367956-2849158
38        403-4367956-2849158
61        404-2262140-4696366
62        404-2262140-4696366
79        408-4069830-3819562
                 ...         
128930    407-8519205-5064368
128941    403-3035549-4142738
128942    403-3035549-4142738
128952    404-8878574-4524358
128953    404-8878574-4524358
Name: Order ID, Length: 15443, dtype: object


**DUPLICATE VALUE OBSERVATIONS**  
ALL ROWS ARE UNIIQUE  
THERE ARE REPEAT ORDER IDs BUT IT MIGHT BE MULTIPE ITEMS PURCHASED DURING THE SAME SHOPPING ORDER  


In [48]:
print(df['Size'].value_counts())

M       22711
L       22132
XL      20876
XXL     18096
S       17090
3XL     14816
XS      11161
6XL       738
5XL       550
4XL       427
Free      378
Name: Size, dtype: int64


**CATEGORY CHECKS**  
Status - MIXED GRANULARITY, REDUNDANCY, CORE STATES (SHIPPED,SHIPPING)  
Fulfillment - *NOTHING*  
Ship-service-level - *NOTHING*  
Sales Channel - *NOTHING*  
Courier Status - Null cells(6872)  
promotion-ids - The column contains multiple, distinct promotion codes separated by commas, Trailing Whitespace, 5,787 unique combinations of promotion IDs, Null cells(49153)  
Category - Inconsitent casing  
Size - *NOTHING*



In [49]:
print(df.isnull().sum())

index                     0
Order ID                  0
Date                      0
Status                    0
Fulfilment                0
Sales Channel             0
ship-service-level        0
Style                     0
SKU                       0
Category                  0
Size                      0
ASIN                      0
Courier Status         6872
Qty                       0
currency               7795
Amount                 7795
ship-city                33
ship-state               33
ship-postal-code         33
ship-country             33
promotion-ids         49153
B2B                       0
fulfilled-by          89698
Unnamed: 22           49050
dtype: int64


In [58]:
placeholders = ['na','--', '_', 'blank', 'null', 'NAN']
placeholder_mask = df['Size'].isin(placeholders).any()
print(placeholder_mask)


False


In [69]:
empty_string = (df['promotion-ids'] == '').any()
print(empty_string)


0         False
1         False
2         False
3         False
4         False
          ...  
128970    False
128971    False
128972    False
128973    False
128974    False
Name: promotion-ids, Length: 128975, dtype: bool


In [82]:
negative_values = df[df['Qty'] <0]
print (len(negative_values))

0


In [81]:
zero_values = df[df['Qty'] ==0]
print (len(zero_values))


12807


**NUMERIC FIELD CHECKS**  
Amount - Has zero values  
Quantity - Has zero vaues


In [83]:
symbol_check = df['Qty'].astype(str).str.contains(r'[₹,\$]')
print({symbol_check.sum()})

{0}


In [86]:
start_date = df['Date'].min()
print (start_date)

03-31-22


In [87]:
start_date = df['Date'].max()
print (start_date)

06-29-22


**DATE RANGE**  
Dates fall into reasonable business period  
No future dates exists  
No year anomalies

In [88]:
standardised_dates = pd.to_datetime(df['Date'], errors = 'coerce')
print (standardised_dates)

0        2022-04-30
1        2022-04-30
2        2022-04-30
3        2022-04-30
4        2022-04-30
            ...    
128970   2022-05-31
128971   2022-05-31
128972   2022-05-31
128973   2022-05-31
128974   2022-05-31
Name: Date, Length: 128975, dtype: datetime64[ns]


In [89]:
inconsistent_dates = standardised_dates.isna()
print (inconsistent_dates)

0         False
1         False
2         False
3         False
4         False
          ...  
128970    False
128971    False
128972    False
128973    False
128974    False
Name: Date, Length: 128975, dtype: bool


In [94]:
inconsistent_entries = df['Date'][inconsistent_dates]
print (len(inconsistent_entries))

0


In [97]:
df['Date'].dtype

dtype('O')

**DATE STORED AS OBJECTS**
